src/ui/Carousel/hooks/index.ts

```diff
@@ -0,0 +1,246 @@
+import { useCallback, useRef, useState } from 'react';
+
+import { useMount, useUnmount, useUpdate, useUpdateOnly } from '@rounik/react-custom-hooks';
+
+import { GlobalModel } from '@services';
+import { RAFIdInfo } from '../../../services';
+
+import { getNextIndex } from '../helpers';
+import { CarouselState, UseCarouselConfig } from '../types';
+
+export const useCarousel = <T>({
+  auto,
+  interval,
+  items,
+  keepDirection,
+  startIndex,
+  styles,
+  toLeft
+}: UseCarouselConfig<T>) => {
+  const [state, setState] = useState<CarouselState<T>>({
+    current: items[startIndex <= items.length - 1 ? startIndex : 0] || null,
+    currentIndex: startIndex <= items.length - 1 ? startIndex : 0,
+    next: null,
+    toLeft,
+    trackStyle: toLeft ? '' : styles.ToRight
+  });
+
+  const transitioning = useRef<boolean>(false);
+
+  const scheduledMove = useRef<'left' | 'right' | null>(null);
+
+  const timeoutId = useRef<RAFIdInfo | null | undefined>(null);
+
+  const move = useCallback(
+    (index?: number) => {
+      if (items.length === 1) {
+        return;
+      }
+
+      transitioning.current = true;
+
+      setState((currentState) => {
+        let current = currentState.current;
+
+        if (currentState.current === null) {
+          current = currentState.next;
+        }
+
+        const nextIndex =
+          typeof index !== 'undefined' && index !== currentState.currentIndex
+            ? index
+            : getNextIndex(currentState.currentIndex, items.length - 1, currentState.toLeft);
+
+        let newCurrentIndex = currentState.currentIndex;
+
+        if (typeof index !== 'undefined' && index !== newCurrentIndex) {
+          newCurrentIndex = currentState.toLeft ? index - 1 : index + 1;
+        }
+
+        return {
+          ...currentState,
+          current,
+          currentIndex: newCurrentIndex,
+          next: items[nextIndex],
+          trackStyle: currentState.toLeft ? styles.MoveToLeft : styles.MoveToRight
+        };
+      });
+    },
+    [items, styles.MoveToLeft, styles.MoveToRight]
+  );
+
+  const startAutoMove = useCallback(
+    (timeout?: number) => {
+      if (auto) {
+        if (timeoutId.current) {
+          GlobalModel.clearRAFTimeout(timeoutId.current);
+        }
+
+        timeoutId.current = GlobalModel.setRAFTimeout(() => move(), timeout ?? interval ?? 0);
+      }
+    },
+    [auto, interval, move]
+  );
+
+  const moveLeft = useCallback(() => {
+    if (!transitioning.current) {
+      setState((currentState) => ({
+        ...currentState,
+        toLeft: true,
+        trackStyle: ''
+      }));
+
+      if (auto) {
+        startAutoMove(50);

```

Comment:
There is "interval" prop for this. These 50 milliseconds are just to override that prop once because here we need to restart the movement "immediately". If no value is provided the "interval" prop is used which may be several seconds.